modules/kvtagger/kvtagger.c

```diff
@@ -0,0 +1,333 @@
+{
+  return (self->default_selector != NULL);
+}
+
+void
+_tag_message(gpointer pmsg, const TagRecord *tag)
+{
+  LogMessage *msg = (LogMessage *) pmsg;
+  log_msg_set_value_by_name(msg, tag->name, tag->value, -1);
+}
+
+static gboolean
+kvtagger_parser_process(LogParser *s, LogMessage **pmsg, const LogPathOptions *path_options,
+                        const gchar *input, gsize input_len)
+{
+  KVTagger *self = (KVTagger *)s;
+  LogMessage *msg = log_msg_make_writable(pmsg, path_options);
+  GString *selector_str = g_string_new(NULL);
+  const gchar *selector = NULL;
+
+  log_template_format(self->selector_template, msg, NULL, LTZ_LOCAL, 0, NULL, selector_str);
+
+  if (kvtagdb_contains(self->tagdb, selector_str->str))
+    selector = selector_str->str;
+  else if (_is_default_selector_set(self))
+    selector = self->default_selector;
+
+  if (selector)
+    kvtagdb_foreach_tag(self->tagdb, selector, _tag_message, (gpointer)msg);
+
+  g_string_free(selector_str, TRUE);
+
+  return TRUE;
+}
+
+static void
+_replace_template(LogTemplate **old_template, LogTemplate *new_template)
+{
+  log_template_unref(*old_template);
+  *old_template = log_template_ref(new_template);
+}
+
+static void
+_replace_tagdb(KVTagDB **old_db, KVTagDB *new_db)
+{
+  kvtagdb_unref(*old_db);
+  *old_db = kvtagdb_ref(new_db);
+}
+
+static LogPipe *
+kvtagger_parser_clone(LogPipe *s)
+{
+  KVTagger *self = (KVTagger *)s;
+  KVTagger *cloned = (KVTagger *)kvtagger_parser_new(s->cfg);
+
+  log_parser_set_template(&cloned->super, log_template_ref(self->super.template));
+  _replace_tagdb(&cloned->tagdb, self->tagdb);
+  _replace_template(&cloned->selector_template, self->selector_template);
+  kvtagger_set_prefix(&cloned->super, self->prefix);
+  kvtagger_set_filename(&cloned->super, self->filename);
+  kvtagger_set_database_default_selector(&cloned->super, self->default_selector);
+
+  return &cloned->super.super;
+}
+
+static void
+kvtagger_parser_free(LogPipe *s)
+{
+  KVTagger *self = (KVTagger *)s;
+
+  kvtagdb_unref(self->tagdb);
+  g_free(self->filename);
+  g_free(self->prefix);
+  log_template_unref(self->selector_template);
+  log_parser_free_method(s);
+  g_free(self->selector_template_string);
+}
+
+static gboolean
+_is_relative_path(const gchar *filename)
+{
+  return (filename[0] != '/');
+}
+
+static gchar *
+_complete_relative_path_with_config_path(const gchar *filename)
+{
+  return g_build_filename(get_installation_path_for(SYSLOG_NG_PATH_SYSCONFDIR), filename, NULL);
+}
+
+static FILE *
+_open_data_file(const gchar *filename)
+{
+  FILE *f = NULL;
+
+  if (_is_relative_path(filename))
+    {
+      gchar *absolute_path = _complete_relative_path_with_config_path(filename);
+      f = fopen(absolute_path, "r");
+      g_free(absolute_path);
+    }
+  else
+    {
+      f = fopen(filename, "r");
+    }
+
+  return f;
+}
+
+const gchar*
+_get_filename_extension(const gchar *filename)
+{
+   gsize filename_len = strlen(filename);
```

Comment:
`strrchr` should be used instead of this hard coding. Some possible edge cases to consider:

    xycsv
    xy.sqlite